File: c++/Easy/Part004/Average Speed.cpp
#include <iostream>
#include <sstream>
#include <iomanip> // For setprecision

using namespace std;

void calculate_average_speed(string input_string) {
    stringstream ss(input_string);
    string pair;
    
    double total_distance = 0.0;
    double total_time = 0.0;
    
    // Read pairs until the end of stringstream
    while (getline(ss, pair, ' ')) {
        // Split the pair into distance and time
        size_t pos = pair.find('@');
        string distance_str = pair.substr(0, pos);
        string time_str = pair.substr(pos + 1);
        
        // Convert distance and time from string to double
        double distance = stod(distance_str);
        double time = stod(time_str);
        
        // Accumulate total distance and total time
        total_distance += distance;
        total_time += time;
    }
    
    // Calculate average speed
    double average_speed = total_distance / total_time;
    
    // Print the result in the required format
    cout << fixed << setprecision(2) << average_speed << " kmph" << endl;
}

int main() {
    string input_string;
    getline(cin, input_string); // Read input dynamically
    
    calculate_average_speed(input_string);
    
    return 0;
}

File: Java/Easy/Part004/Average Speed.java
import java.util.Scanner;

public class Hello {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        // Read the input line
        String line = scanner.nextLine();
        
        // Initialize total distance and total time
        double totalDistance = 0.0;
        double totalTime = 0.0;
        
        // Split the input line by spaces to process each DISTANCE@TIMETAKEN
        String[] segments = line.split(" ");
        
        for (String segment : segments) {
            // Split each segment by '@' to separate distance and time
            String[] parts = segment.split("@");
            if (parts.length == 2) {
                try {
                    // Parse distance and time
                    double distance = Double.parseDouble(parts[0]);
                    double time = Double.parseDouble(parts[1]);
                    
                    // Accumulate total distance and total time
                    totalDistance += distance;
                    totalTime += time;
                } catch (NumberFormatException e) {
                    // Handle the case where distance or time is not a valid number
                    System.err.println("Invalid input format");
                    return;
                }
            } else {
                // Handle the case where the segment does not contain exactly one '@'
                System.err.println("Invalid input format");
                return;
            }
        }
        
        // Calculate the average speed
        if (totalTime != 0) {
            double averageSpeed = totalDistance / totalTime;
            
            // Print the average speed formatted to 2 decimal places
            System.out.printf("%.2f kmph%n", averageSpeed);
        } else {
            // Handle the case where total time is 0 (should not happen in valid inputs)
            System.err.println("Total time cannot be zero");
        }
    }
}
